myriad-scheduler/src/main/java/com/ebay/myriad/scheduler/JHSTaskFactoryImpl.java

```diff
@@ -0,0 +1,351 @@
+        } else {
+          String address = conf.get(YARN_RESOURCEMANAGER_WEBAPP_ADDRESS);
+          if (address == null || address.isEmpty()) {
+            address = conf.get(YARN_RESOURCEMANAGER_HOSTNAME) + ":8088";
+          }
+          return "http://" + address + "/conf";
+        }
+      }
+
+      /**
+       * Helper method to reserve ports
+       * @param offer
+       * @param requestedPorts
+       * @return
+       */
+      private List<Long> getAvailablePorts(Offer offer, int requestedPorts) {
+        if (requestedPorts == 0) {
+          return null;
+        }
+        final List<Long> returnedPorts = new ArrayList<>();
+        for (Resource resource : offer.getResourcesList()){
+          if (resource.getName().equals("ports")){
+            Iterator<Value.Range> itr = resource.getRanges().getRangeList().iterator();
+            while (itr.hasNext()) {
+              Value.Range range = itr.next();
+              if (range.getBegin() <= range.getEnd()) {
+                long i = range.getBegin();
+                while (i <= range.getEnd() && returnedPorts.size() < requestedPorts) {
+                  returnedPorts.add(i);
+                  i++;
+                }
+                if (returnedPorts.size() >= requestedPorts) { 
+                  return returnedPorts;
+                }
+              }
+            }
+          }
+        }
+        // this is actually an error condition - we did not have enough ports to use
+        return returnedPorts;
+      }
+
+      @Override
+      public ExecutorInfo getExecutorInfoForSlave(FrameworkID frameworkId, Offer offer,
+          CommandInfo commandInfo) {
+        // nothing to implement here, since we are using default slave executor
+        return null;
+      }
+      @Override
+      public TaskConstraints getConstraints() {
+        return constraints;
+      }
+      
+      
+      /**
+       * Implement NM Task Constraints
+       *
+       */
+      static class JHSTaskConstraints implements TaskConstraints {
+
+        private int portsCount;
+        
+        JHSTaskConstraints(MyriadConfiguration cfg) {
+          this.portsCount = 0;
+          Map<String, AuxTaskConfiguration> auxConfigs = cfg.getAuxTaskConfigurations();
+          if (auxConfigs == null) {
+            return;
+          }
+          // it is overkill, but just in case name of the service is different from "jobhistory"
+          for (Map.Entry<String, AuxTaskConfiguration> auxConfig : auxConfigs.entrySet()) {
+            final AuxTaskConfiguration auxCfg = auxConfig.getValue();
+            if (JHSTaskFactoryImpl.class.getName().equals(auxCfg.getTaskFactoryImplName())) {
+              if (auxCfg.getPorts() != null) {
+                this.portsCount = auxCfg.getPorts().size();
+              }
+              return;
+            }
+          }
+        }
+        
+        @Override
+        public int portsCount() {
+          return portsCount;
+        }
+      }
+    }
```

Comment:
Lots of this code is redundant with NMExecutorCLGenImpl.java and DownloadNMExecutorCLGenImpl.java. Seems like we should get the right abstraction for code reuse here.